sqld/proto/wal_log.proto

```diff
@@ -2,21 +2,26 @@ syntax = "proto3";
 package wal_log;
 
 message LogOffset {
-    uint64 startOffset = 1;
+    // offset of the last reveiced frame. If absent, the server should send frame 0.
+    optional uint64 currentOffset = 1;
```

Comment:
This is not backwards compatible, is it? If a primary with a previous version receives an empty LogOffset, it will not know how to handle it, will it?

Reply:
no it's not backward compatible

Reply:
Can we make it backward compatible? We should avoid breaking changes like this when not absolutely necessary.

Reply:
the protocol is not backward compatible, since it now requires a handshake

Reply:
Yes but this change makes it impossible to do rolling upgrade of the cluster.
Normally you would update replicas one by one and then update the primary - no downtime for reads, short downtime for writes.

It's a good practise to not require the whole cluster to be turned down for upgrade.

Reply:
You have a plan for rolling upgrades with only the handshake?

Reply:
Yes. Were you planning to shut everything down for update?

Reply:
So we leave databases for existing ~100 users broken?

The way I envision it:
1. We don't break proto compatibility
2. New version of the replica should gracefully handle if hello message is rejected and assume everything is good in such case
3. We upgrade all replicas and then switch the primary

I think that would work, no?

Reply:
let's consider the two scenarios:
1. we first update the primary and then replicas:
If we do that, the primary is going to return `NO_HELLO` when old replicas try to replicate. We could make the primary ignore when the replica hasn't made a handshake, but how do you differentiate an old replica that does not know about hello, and a new one that hasn't performed a handshake? This makes the handshake pointless
2. we upgrade replicas first and then primary:
replicas will send a handshake to the server and the server will return an error. What you suggest is to proceed with asking for frames anyway, but that kills the purpose of the handshake again. The purpose of the handshake is for the replica and the primary to agree on the data that will be exchanged.

Reply:
So I was thinking about (2). I don't think it defeats the purpose of the handshake. If the primary returns error on the handshake then we assume it's the old primary that does not understand the hello message. As long as we distinguish errors returned from the new primary from "don't understand the message" error, it all should be good and the replica should be able to decide if it deals with an old primary or new primary with error.

Reply:
In the current state, the primary cannot be upgraded. The database directory structure has been changed in #94, the wal format has been changed in #72 on which this PR is based, along with the format of the frame being sent, and other changes to the protocol. We are working on protocol improvements because we can't even assume that restarting the primary is safe at this point. 

Moreover, I thought we agreed that breaking was fine at his point during the previous town hall, and it is under this assumption that #94 was merged.

Reply:
My understanding is that we can break things when there's no other way and when it gives us in return something valuable.

We should start thinking about doing changes in such a way that would allow us to upgrade existing clusters. We have ~100 users now and loosing even few of them because of not being able to fix their clusters is a very very bad thing.

I'm particularly unhappy with the fact that we have existing clusters that have broken replication and now we won't be able to fix them.

@penberg my understanding was that we start doing changes in a more responsible way with at least a migration path. Has anything changed?

Reply:
So https://github.com/libsql/sqld/pull/94 shouldn't prevent us from upgrading the primary - can't we just move the db file to the new dir and start the primary in a new version?

Reply:
https://github.com/libsql/sqld/pull/72 is not yet merged so maybe we should first fix the replication and only then merge https://github.com/libsql/sqld/pull/72. Is this PR using anything but database_id from https://github.com/libsql/sqld/pull/72 @MarinPostma ?

Reply:
It seems that database_id check is not crucial for replication so we can:
1. fix replication
2. do https://github.com/libsql/sqld/pull/72
3. Add database_id check to handshake

no?

I think it is important to fix existing clusters and not just ignore them if we can and it seems that we can.

Reply:
we can, we just need to do it - in the same way we could manually dump all existing cluster and re-import them in the new version

Reply:
it sucks but maybe that's what we should do. That brings us back to the backwards compatibility issue

Reply:
Resolving this as we agreed on the dump - load migration procedure.